File: ecs_microfinance/doctype_triggers/hr/staffing_plan/staffing_plan.py
from __future__ import unicode_literals
import frappe
from frappe import _


@frappe.whitelist()
def before_insert(doc, method=None):
    pass
@frappe.whitelist()
def after_insert(doc, method=None):
    pass
@frappe.whitelist()
def onload(doc, method=None):
    pass
@frappe.whitelist()
def before_validate(doc, method=None):
    pass
@frappe.whitelist()
def validate(doc, method=None):
    pass
@frappe.whitelist()
def on_submit(doc, method=None):
    pass
@frappe.whitelist()
def on_cancel(doc, method=None):
    pass
@frappe.whitelist()
def on_update_after_submit(doc, method=None):
    pass
@frappe.whitelist()
def before_save(doc, method=None):
    pass
@frappe.whitelist()
def before_cancel(doc, method=None):
    pass
@frappe.whitelist()
def on_update(doc, method=None):
    pass


# @frappe.whitelist()
# def get_job_request(doc,from_date,to_date, method=None):

#     emps = frappe.db.sql(""" select a.designation, a.no_of_positions, a.expected_salary, b.name,b.posting_date
#                                                                from `tabJob Request Detail` a join `tabJob Request` b
#                                                                on a.parent = b.name
#                                                                where b.posting_date >='{from1}'
#                                                                and b.posting_date <='{to}'

#                                                            """.format(from1=from_date,
#                                                                       to=to_date), as_dict=1)
#     if emps:
        # event = frappe.get_doc('Staffing Plan', doc)
        # event.staffing_details = []
        #
        # for y in emps:
        #     items = event.append("staffing_details", {})
        #     items.employee = y.employee
        #     items.employee_name = y.employee_name
        #     items.designation = y.designation
        #     items.no_of_positions = y.vacancies
        #     items.expected_salary = y.estimated_cost_per_position
    #     return emps

    # else:
    #     frappe.throw("لا يوجد موظفين في هذه الفترة")



# @frappe.whitelist()
# def set_job_requisitions(designation):
#     requisitions = frappe.db.get_list(
#         "Job Requisition",
#         filters={"designation": designation},
#         fields=["name","designation", "no_of_positions", "expected_compensation","custom_branch","custom_branch","custom_basic_salary","custom_incentive","custom_allowances"],
#     )
#     total = 0.0
#     previous_staffing_plan = [parent['parent'] for parent in frappe.db.get_all("Staffing Plan Detail", {"custom_job_requisition": requisitions["name"]}, ["parent"])]
#     if previous_staffing_plan:
#         previous_staffing_plan_name =[name['name'] for name in frappe.db.get_all("Staffing Plan", {"name": ["in", previous_staffing_plan], "docstatus": 1}, ["name"])]
#         # frappe.msgprint("previous_staffing_plan_name:"+str(previous_staffing_plan_name))
#         previous_vacancies = frappe.db.get_all("Staffing Plan Detail", {"custom_job_requisition":requisitions["name"], "parent": ["in", previous_staffing_plan_name]}, ["vacancies"])
#         # frappe.msgprint("previous_custom_total_planned_manpower:"+str(previous_custom_total_planned_manpower))
#         for row in previous_vacancies:
#             total += row["vacancies"]

#         requisitions["total"] = total

#         # self.append(
#         #     "staffing_details",
#         #     {
#         #         "designation": req.designation,
#         #         "custom_job_requisition":req.name,
#         #         "vacancies": req.no_of_positions,
#         #         "estimated_cost_per_position": req.expected_compensation,
#         #         "custom_previous_planned_manpower": total,
#         #         "custom_branch":req.custom_branch,
#         #         "custom_basic_salary":req.custom_basic_salary,
#         #         "custom_incentive":req.custom_incentive,
#         #         "custom_allowances":req.custom_allowances
#         #     },
#         # )

#     return requisitions
        

# @frappe.whitelist()
# def set_job_requisitions(designation):
#     # Retrieve job requisitions based on the provided designation
#     requisitions = frappe.db.get_list(
#         "Job Requisition",
#         filters={"designation": designation},
#         fields=["name", "designation", "no_of_positions", "expected_compensation", "custom_branch", "custom_basic_salary", "custom_incentive", "custom_allowances"]
#     )

#     # Calculate total vacancies from previous staffing plans
#     # total = 0.0
#     # previous_staffing_plan = [parent['parent'] for parent in frappe.db.get_all("Staffing Plan Detail", {"custom_job_requisition": requisitions[0]["name"]}, ["parent"])]
#     # if previous_staffing_plan:
#     #     previous_staffing_plan_name = [name['name'] for name in frappe.db.get_all("Staffing Plan", {"name": ["in", previous_staffing_plan], "docstatus": 1}, ["name"])]
#     #     previous_vacancies = frappe.db.get_all("Staffing Plan Detail", {"custom_job_requisition": requisitions[0]["name"], "parent": ["in", previous_staffing_plan_name]}, ["vacancies"])
#     #     for row in previous_vacancies:
#     #         total += row["vacancies"]
        
#     #     requisitions["total"] = total

#     return requisitions
        
@frappe.whitelist()
def get_salary(designation,branch=None):
    designation_doc = frappe.get_value("Designation",{"name":designation},["*"],as_dict = 1)
    # Calculate total vacancies from previous staffing plans
    total = 0.0
    if branch:
        previous_staffing_plan = [parent['parent'] for parent in frappe.db.get_all("Staffing Plan Detail", {"designation":designation,"custom_branch":branch}, ["parent"])]
    else:
        previous_staffing_plan = [parent['parent'] for parent in frappe.db.get_all("Staffing Plan Detail", {"designation":designation,"custom_branch":""}, ["parent"])]

    if previous_staffing_plan:
        previous_staffing_plan_name = [name['name'] for name in frappe.db.get_all("Staffing Plan", {"name": ["in", previous_staffing_plan], "docstatus": 1}, ["name"])]
        previous_vacancies = frappe.db.get_all("Staffing Plan Detail", {"parent": ["in", previous_staffing_plan_name]}, ["vacancies"])
        for row in previous_vacancies:
            total += row["vacancies"]
        designation_doc["total"] = total
    return designation_doc

@frappe.whitelist()
def get_previous_planned(designation,branch = None):
    # Calculate total vacancies from previous staffing plans
    total = 0.0
    if branch:
        previous_staffing_plan = [parent['parent'] for parent in frappe.db.get_all("Staffing Plan Detail", {"designation":designation,"custom_branch":branch}, ["parent"])]
    else:
        previous_staffing_plan = [parent['parent'] for parent in frappe.db.get_all("Staffing Plan Detail", {"designation":designation,"custom_branch":""}, ["parent"])]

    if previous_staffing_plan:
        previous_staffing_plan_name = [name['name'] for name in frappe.db.get_all("Staffing Plan", {"name": ["in", previous_staffing_plan], "docstatus": 1}, ["name"])]
        previous_vacancies = frappe.db.get_all("Staffing Plan Detail", {"parent": ["in", previous_staffing_plan_name]}, ["vacancies"])
        for row in previous_vacancies:
            total += row["vacancies"]

    return total

File: ecs_microfinance/doctype_triggers/hr/staffing_plan/staffing_plan.js
//// Copyright (c) 2023, erpcloudsystems and contributors
//// For license information, please see license.txt
//
//frappe.ui.form.on('Staffing Plan', {
//	get_job_request: function(frm) {
//		frm.doc.cows = []
//		frappe.call({
//			doc: frm.doc,
//			method: "get_job_request",
//			callback: function(r) {
//				frm.refresh_fields();
//				frm.refresh();
//			}
//		});
//	}
//});

// frappe.ui.form.on('Staffing Plan Detail', {
//     designation: function(frm, cdt, cdn) {
//         let child = locals[cdt][cdn];
//         console.log("kkkk");
//         if (child.designation){
//         frappe.call({
//             method: "ecs_microfinance.doctype_triggers.hr.staffing_plan.staffing_plan.set_job_requisitions",
//             args: {
//                 designation: child.designation,
//             },
//             callback: function(r) {
//                 if (r.message) {
//                     console.log(r.message);
//                     // frm.set_value('custom_job_requisition', r.message.name);
//                     frappe.model.set_value(cdt,cdn,'vacancies', r.message.no_of_positions);
//                     frappe.model.set_value(cdt.cdn,'estimated_cost_per_position', r.message.expected_compensation);
//                     // frm.set_value('custom_previous_planned_manpower', r.message.total);
//                     frappe.model.set_value(cdt,cdn,'custom_branch', r.message.custom_branch);
//                     frappe.model.set_value(cdt,cdn,'custom_basic_salary', r.message.custom_basic_salary);
//                     frappe.model.set_value(cdt,cdn,'custom_incentive', r.message.custom_incentive);
//                     frappe.model.set_value(cdt,cdn,'custom_allowances', r.message.custom_allowances );
//                  }
            
//                     }
//         });
//         frm.refresh_fields(["custom_job_requisition", "vacancies", "estimated_cost_per_position", "estimated_cost_per_position"
//     ,"custom_previous_planned_manpower","custom_branch","custom_basic_salary","custom_incentive","custom_allowances"]);
//     }}
// });

frappe.ui.form.on('Staffing Plan Detail',{
    designation: function(frm, cdt, cdn) {
        let child = locals[cdt][cdn];
    if (child.designation){
    frappe.call({
        method: "ecs_microfinance.doctype_triggers.hr.staffing_plan.staffing_plan.get_salary",
        args: {
            designation: child.designation,
            branch: child.custom_branch
        },
        callback: function(r) {
            // Handle the response here if necessary
            console.log(r.message);
            var total_allowances = r.message.custom_other_allowance + r.message.custom_mobile_allowance + r.message.custom_living_allowance + r.message.housing_allowance + r.message.transportation_allowance
            frappe.model.set_value(cdt,cdn,'custom_basic_salary', r.message.basic_salary);
            frappe.model.set_value(cdt,cdn,'custom_allowances',total_allowances);
            frappe.model.set_value(cdt,cdn,'custom_incentive',r.message.custom_incentives);
            frappe.model.set_value(cdt,cdn,'custom_previous_planned_manpower',r.message.total);
            frappe.model.set_value(cdt,cdn,'estimated_cost_per_position',r.message.basic_salary + total_allowances +  r.message.custom_incentives);
        //     frm.set_value('expected_compensation', r.message.basic_salary + total_allowances +  r.message.custom_incentives);
        }
    });}
    }
});

frappe.ui.form.on('Staffing Plan',{
    refresh: function(frm) {
        cur_frm.fields_dict['staffing_details'].grid.get_field("designation").get_query = function(doc, cdt, cdn) {
            return {
                filters: [
                    ['Designation', 'department', '=',frm.doc.department],
                ]
            };
        };
    }
});


frappe.ui.form.on('Staffing Plan',{
    refresh: function(frm) {
        cur_frm.fields_dict['staffing_details'].grid.get_field("custom_branch").get_query = function(doc, cdt, cdn) {
            return {
                filters: [
                    ['Branch', 'custom_governorate', '=',frm.doc.custom_governorate],
                ]
            };
        };
    }
});


frappe.ui.form.on('Staffing Plan Detail',{
    custom_branch:  function(frm, cdt, cdn){
        let child = locals[cdt][cdn];
        if (child.designation){
            frappe.call({
                method: "ecs_microfinance.doctype_triggers.hr.staffing_plan.staffing_plan.get_previous_planned",
                args: {
                    designation: child.designation,
                    branch: child.custom_branch
                },
                callback:function(r){
                    frappe.model.set_value(cdt,cdn,'custom_previous_planned_manpower',r.message);
                }

            })

    }}
})

frappe.ui.form.on('Staffing Plan Detail',{
    custom_others:  function(frm, cdt, cdn){
        let child = locals[cdt][cdn];
        if (child.estimated_cost_per_position){
        frappe.model.set_value(cdt,cdn,'estimated_cost_per_position',child.custom_incentive + child.custom_allowances +  child.custom_basic_salary + child.custom_others);
        }
        else{
        frappe.model.set_value(cdt,cdn,'estimated_cost_per_position',child.custom_incentive + child.custom_allowances +  child.custom_basic_salary);
        }

    }
    })